File: packages/google_mlkit_genai_prompt/android/src/main/java/com/google_mlkit_genai_prompt/Prompt.java
package com.google_mlkit_genai_prompt;

import android.content.Context;

import androidx.annotation.NonNull;

import com.google.common.util.concurrent.Futures;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.common.util.concurrent.FutureCallback;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class Prompt implements MethodChannel.MethodCallHandler {
    private static final String CHECK_FEATURE_STATUS = "genai#checkFeatureStatus";
    private static final String DOWNLOAD_FEATURE = "genai#downloadFeature";
    private static final String RUN_INFERENCE = "genai#runInference";
    private static final String RUN_INFERENCE_STREAMING = "genai#runInferenceStreaming";
    private static final String CLOSE = "genai#closePrompt";

    private final Context context;
    private final Map<String, Object> instances = new HashMap<>();
    private final Executor executor = Executors.newSingleThreadExecutor();

    public Prompt(Context context) {
        this.context = context;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        String method = call.method;
        switch (method) {
            case CHECK_FEATURE_STATUS:
                checkFeatureStatus(call, result);
                break;
            case DOWNLOAD_FEATURE:
                downloadFeature(call, result);
                break;
            case RUN_INFERENCE:
                runInference(call, result);
                break;
            case RUN_INFERENCE_STREAMING:
                runInferenceStreaming(call, result);
                break;
            case CLOSE:
                closePrompt(call);
                result.success(null);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private Object initialize(MethodCall call) {
        // Prompt API initialization - structure may vary
        // This is a placeholder implementation
        try {
            // Try to use Generation.INSTANCE.getClient() for Java
            Object generationInstance = Class.forName("com.google.mlkit.genai.prompt.Generation").getField("INSTANCE").get(null);
            Object generativeModel = generationInstance.getClass().getMethod("getClient").invoke(generationInstance);
            Object generativeModelFutures = Class.forName("com.google.mlkit.genai.prompt.GenerativeModelFutures")
                .getMethod("from", Class.forName("com.google.mlkit.genai.prompt.GenerativeModel"))
                .invoke(null, generativeModel);
            return generativeModelFutures;
        } catch (Exception e) {
            // If reflection fails, return a placeholder
            return new Object();
        }
    }

    private void checkFeatureStatus(MethodCall call, MethodChannel.Result result) {
        String id = call.argument("id");
        Object generativeModel = instances.get(id);
        if (generativeModel == null) {
            generativeModel = initialize(call);
            instances.put(id, generativeModel);
        }

        try {
            ListenableFuture<Integer> future = (ListenableFuture<Integer>) generativeModel.getClass()
                .getMethod("checkStatus").invoke(generativeModel);
            Futures.addCallback(future, new FutureCallback<Integer>() {
                @Override
                public void onSuccess(Integer status) {
                    int statusValue;
                    if (status == com.google.mlkit.genai.common.FeatureStatus.UNAVAILABLE) {
                        statusValue = 0;
                    } else if (status == com.google.mlkit.genai.common.FeatureStatus.DOWNLOADABLE) {
                        statusValue = 1;
                    } else if (status == com.google.mlkit.genai.common.FeatureStatus.DOWNLOADING) {
                        statusValue = 2;
                    } else if (status == com.google.mlkit.genai.common.FeatureStatus.AVAILABLE) {
                        statusValue = 3;
                    } else {
                        statusValue = 0;
                    }
                    result.success(statusValue);
                }

                @Override
                public void onFailure(Throwable e) {
                    result.error("PromptError", e.toString(), null);
                }
            }, executor);
        } catch (Exception e) {
            result.error("PromptError", "Failed to check status: " + e.toString(), null);
        }
    }

    private void downloadFeature(MethodCall call, MethodChannel.Result result) {
        String id = call.argument("id");
        Object generativeModel = instances.get(id);
        if (generativeModel == null) {
            generativeModel = initialize(call);
            instances.put(id, generativeModel);
        }

        try {
            generativeModel.getClass().getMethod("download", com.google.mlkit.genai.common.DownloadCallback.class)
                .invoke(generativeModel, new com.google.mlkit.genai.common.DownloadCallback() {
                    @Override
                    public void onDownloadStarted(long bytesToDownload) {
                        // Handle download started
                    }

                    @Override
                    public void onDownloadFailed(com.google.mlkit.genai.common.GenAiException e) {
                        result.error("DownloadError", e.toString(), null);
                    }

                    @Override
                    public void onDownloadProgress(long totalBytesDownloaded) {
                        // Handle download progress
                    }

                    @Override
                    public void onDownloadCompleted() {
                        result.success(null);
                    }
                });
        } catch (Exception e) {
            result.error("PromptError", "Failed to download: " + e.toString(), null);
        }
    }

    private void runInference(MethodCall call, MethodChannel.Result result) {
        // Prompt API inference - placeholder implementation
        result.error("PromptError", "Prompt API inference not yet fully implemented", null);
    }

    private void runInferenceStreaming(MethodCall call, MethodChannel.Result result) {
        // Streaming implementation would require EventChannel
        result.notImplemented();
    }

    private void closePrompt(MethodCall call) {
        String id = call.argument("id");
        instances.remove(id);
    }
}

File: packages/google_mlkit_genai_prompt/android/src/main/kotlin/com/google_mlkit_genai_prompt/Prompt.kt
package com.google_mlkit_genai_prompt

import android.content.Context
import com.google.common.util.concurrent.FutureCallback
import com.google.common.util.concurrent.Futures
import com.google.common.util.concurrent.ListenableFuture
import com.google.mlkit.genai.common.DownloadCallback
import com.google.mlkit.genai.common.FeatureStatus
import com.google.mlkit.genai.common.GenAiException
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.concurrent.Executor
import java.util.concurrent.Executors

class Prompt(
    private val context: Context,
) : MethodChannel.MethodCallHandler {
    private val instances = mutableMapOf<String, Any>()

    companion object {
        private const val CHECK_FEATURE_STATUS = "genai#checkFeatureStatus"
        private const val DOWNLOAD_FEATURE = "genai#downloadFeature"
        private const val RUN_INFERENCE = "genai#runInference"
        private const val RUN_INFERENCE_STREAMING = "genai#runInferenceStreaming"
        private const val CLOSE = "genai#closePrompt"
    }

    private val executor: Executor = Executors.newSingleThreadExecutor()

    override fun onMethodCall(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        when (call.method) {
            CHECK_FEATURE_STATUS -> {
                checkFeatureStatus(call, result)
            }

            DOWNLOAD_FEATURE -> {
                downloadFeature(call, result)
            }

            RUN_INFERENCE -> {
                runInference(call, result)
            }

            RUN_INFERENCE_STREAMING -> {
                result.notImplemented()
            }

            CLOSE -> {
                closePrompt(call)
                result.success(null)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    private fun initialize(call: MethodCall): Any =
        try {
            val generationInstance =
                Class
                    .forName("com.google.mlkit.genai.prompt.Generation")
                    .getField("INSTANCE")
                    .get(null)
            val generativeModel =
                generationInstance
                    ?.javaClass
                    ?.getMethod("getClient")
                    ?.invoke(generationInstance)
            Class
                .forName("com.google.mlkit.genai.prompt.GenerativeModelFutures")
                .getMethod("from", Class.forName("com.google.mlkit.genai.prompt.GenerativeModel"))
                .invoke(null, generativeModel) ?: Any()
        } catch (e: Exception) {
            Any()
        }

    private fun checkFeatureStatus(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val id = call.argument<String>("id") ?: return result.error("PromptError", "Missing id", null)
        val generativeModel = instances.getOrPut(id) { initialize(call) }

        try {
            @Suppress("UNCHECKED_CAST")
            val future =
                generativeModel.javaClass
                    .getMethod("checkStatus")
                    .invoke(generativeModel) as ListenableFuture<Int>

            Futures.addCallback(
                future,
                object : FutureCallback<Int> {
                    override fun onSuccess(status: Int?) {
                        val statusValue =
                            when (status) {
                                FeatureStatus.UNAVAILABLE -> 0
                                FeatureStatus.DOWNLOADABLE -> 1
                                FeatureStatus.DOWNLOADING -> 2
                                FeatureStatus.AVAILABLE -> 3
                                else -> 0
                            }
                        result.success(statusValue)
                    }

                    override fun onFailure(e: Throwable) {
                        result.error("PromptError", e.toString(), null)
                    }
                },
                executor,
            )
        } catch (e: Exception) {
            result.error("PromptError", "Failed to check status: $e", null)
        }
    }

    private fun downloadFeature(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val id = call.argument<String>("id") ?: return result.error("PromptError", "Missing id", null)
        val generativeModel = instances.getOrPut(id) { initialize(call) }

        try {
            generativeModel.javaClass
                .getMethod("download", DownloadCallback::class.java)
                .invoke(
                    generativeModel,
                    object : DownloadCallback {
                        override fun onDownloadStarted(p0: Long) {}

                        override fun onDownloadFailed(e: GenAiException) {
                            result.error("DownloadError", e.toString(), null)
                        }

                        override fun onDownloadProgress(totalBytesDownloaded: Long) {}

                        override fun onDownloadCompleted() {
                            result.success(null)
                        }
                    },
                )
        } catch (e: Exception) {
            result.error("PromptError", "Failed to download $e", null)
        }
    }

    private fun runInference(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        result.error("PromptError", "Prompt API inference not yet fully implemented", null)
    }

    private fun closePrompt(call: MethodCall) {
        val id = call.argument<String>("id") ?: return
        instances.remove(id)
    }
}
